#!/usr/bin/env python3
"""
–¢–µ—Å—Ç–æ–≤—ã–π —Å–∫—Ä–∏–ø—Ç –¥–ª—è –ø—Ä–æ–≤–µ—Ä–∫–∏ gRPC –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è –∫ —É–¥–∞–ª—ë–Ω–Ω–æ–º—É —Å–µ—Ä–≤–µ—Ä—É
"""
import asyncio
import grpc
from grpc import aio
import sys
import os

# –î–æ–±–∞–≤–ª—è–µ–º –ø—É—Ç—å –∫ protobuf —Ñ–∞–π–ª–∞–º
sys.path.append(os.path.join(os.path.dirname(__file__), '../modules/grpc_service'))
try:
    import streaming_pb2
    import streaming_pb2_grpc
except ImportError as e:
    print(f"‚ùå –û—à–∏–±–∫–∞ –∏–º–ø–æ—Ä—Ç–∞ protobuf: {e}")
    print("–£–±–µ–¥–∏—Ç–µ—Å—å, —á—Ç–æ protobuf —Ñ–∞–π–ª—ã —Å–≥–µ–Ω–µ—Ä–∏—Ä–æ–≤–∞–Ω—ã:")
    print("  cd server/modules/grpc_service")
    print("  python -m grpc_tools.protoc -I. --python_out=. --grpc_python_out=. streaming.proto")
    sys.exit(1)

async def test_grpc_connection(host: str = "20.63.24.187", port: int = 443):
    """–¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ gRPC –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è"""
    address = f"{host}:{port}"
    print(f"üîç –¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ gRPC –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è –∫ {address}...")
    print()
    
    try:
        # –î–ª—è –ø–æ—Ä—Ç–∞ 443 –∏—Å–ø–æ–ª—å–∑—É–µ–º secure_channel —Å TLS
        # –í–ê–ñ–ù–û: insecure_channel –ù–ï —Ä–∞–±–æ—Ç–∞–µ—Ç —Å TLS –ø–æ—Ä—Ç–æ–º 443
        # –î–ª—è self-signed —Å–µ—Ä—Ç–∏—Ñ–∏–∫–∞—Ç–∞ –Ω—É–∂–Ω–æ –ª–∏–±–æ —Å–∫–∞—á–∞—Ç—å —Å–µ—Ä—Ç–∏—Ñ–∏–∫–∞—Ç,
        # –ª–∏–±–æ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å —ç–∫—Å–ø–µ—Ä–∏–º–µ–Ω—Ç–∞–ª—å–Ω—ã–π API
        if port == 443:
            # –ü—Ä–æ–±—É–µ–º —Å–∫–∞—á–∞—Ç—å —Å–µ—Ä—Ç–∏—Ñ–∏–∫–∞—Ç —Å–µ—Ä–≤–µ—Ä–∞ –¥–ª—è –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏—è –≤ credentials
            import subprocess
            import tempfile
            import ssl
            
            try:
                # –°–∫–∞—á–∏–≤–∞–µ–º —Å–µ—Ä—Ç–∏—Ñ–∏–∫–∞—Ç —Å–µ—Ä–≤–µ—Ä–∞
                cert_pem = None
                try:
                    result = subprocess.run(
                        ['openssl', 's_client', '-connect', address, '-showcerts'],
                        input=b'', capture_output=True, timeout=5, check=False
                    )
                    if result.returncode == 0:
                        # –ò–∑–≤–ª–µ–∫–∞–µ–º —Å–µ—Ä—Ç–∏—Ñ–∏–∫–∞—Ç –∏–∑ –≤—ã–≤–æ–¥–∞
                        cert_start = result.stdout.find(b'-----BEGIN CERTIFICATE-----')
                        if cert_start != -1:
                            cert_end = result.stdout.find(b'-----END CERTIFICATE-----', cert_start)
                            if cert_end != -1:
                                cert_pem = result.stdout[cert_start:cert_end + len(b'-----END CERTIFICATE-----')]
                except Exception:
                    pass
                
                if cert_pem:
                    # –ò—Å–ø–æ–ª—å–∑—É–µ–º —Å–∫–∞—á–∞–Ω–Ω—ã–π —Å–µ—Ä—Ç–∏—Ñ–∏–∫–∞—Ç
                    credentials = grpc.ssl_channel_credentials(root_certificates=cert_pem)
                    print("üîí –ò—Å–ø–æ–ª—å–∑—É–µ—Ç—Å—è secure_channel —Å —Å–µ—Ä—Ç–∏—Ñ–∏–∫–∞—Ç–æ–º —Å–µ—Ä–≤–µ—Ä–∞")
                else:
                    # Fallback: –∏—Å–ø–æ–ª—å–∑—É–µ–º —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—ã–µ credentials
                    # –í production —ç—Ç–æ –¥–æ–ª–∂–Ω–æ —Ä–∞–±–æ—Ç–∞—Ç—å —Å –ø—Ä–∞–≤–∏–ª—å–Ω—ã–º —Å–µ—Ä—Ç–∏—Ñ–∏–∫–∞—Ç–æ–º
                    credentials = grpc.ssl_channel_credentials()
                    print("üîí –ò—Å–ø–æ–ª—å–∑—É–µ—Ç—Å—è secure_channel (—Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—ã–µ credentials)")
                    print("‚ö†Ô∏è  –î–ª—è self-signed –º–æ–∂–µ—Ç –ø–æ—Ç—Ä–µ–±–æ–≤–∞—Ç—å—Å—è —Å–∫–∞—á–∞—Ç—å —Å–µ—Ä—Ç–∏—Ñ–∏–∫–∞—Ç")
                
                channel = aio.secure_channel(address, credentials)
            except Exception as e:
                print(f"‚ö†Ô∏è  –û—à–∏–±–∫–∞ —Å–æ–∑–¥–∞–Ω–∏—è secure_channel: {e}")
                print("   –ü—Ä–æ–±—É–µ–º insecure_channel (–º–æ–∂–µ—Ç –Ω–µ —Ä–∞–±–æ—Ç–∞—Ç—å —Å TLS –ø–æ—Ä—Ç–æ–º)...")
                channel = aio.insecure_channel(address)
        else:
            # –î–ª—è –ª–æ–∫–∞–ª—å–Ω–æ–≥–æ —Ç–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏—è –∏—Å–ø–æ–ª—å–∑—É–µ–º insecure channel
            channel = aio.insecure_channel(address)
            print("üîì –ò—Å–ø–æ–ª—å–∑—É–µ—Ç—Å—è insecure_channel (–ª–æ–∫–∞–ª—å–Ω–æ–µ —Ç–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ)")
        
        stub = streaming_pb2_grpc.StreamingServiceStub(channel)
        
        print("‚úÖ –ö–∞–Ω–∞–ª —Å–æ–∑–¥–∞–Ω")
        
        # –ü—Ä–æ–≤–µ—Ä—è–µ–º –ø–æ–¥–∫–ª—é—á–µ–Ω–∏–µ
        try:
            await asyncio.wait_for(channel.channel_ready(), timeout=10.0)
            print("‚úÖ –ö–∞–Ω–∞–ª –≥–æ—Ç–æ–≤ –∫ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏—é")
        except asyncio.TimeoutError:
            print("‚ö†Ô∏è –¢–∞–π–º–∞—É—Ç –æ–∂–∏–¥–∞–Ω–∏—è –≥–æ—Ç–æ–≤–Ω–æ—Å—Ç–∏ –∫–∞–Ω–∞–ª–∞ (10 —Å–µ–∫)")
        except Exception as e:
            print(f"‚ö†Ô∏è –û—à–∏–±–∫–∞ –≥–æ—Ç–æ–≤–Ω–æ—Å—Ç–∏ –∫–∞–Ω–∞–ª–∞: {e}")
        
        # –ü—Ä–æ–±—É–µ–º –æ—Ç–ø—Ä–∞–≤–∏—Ç—å –ø—Ä–æ—Å—Ç–æ–π –∑–∞–ø—Ä–æ—Å
        print()
        print("üì§ –û—Ç–ø—Ä–∞–≤–∫–∞ —Ç–µ—Å—Ç–æ–≤–æ–≥–æ –∑–∞–ø—Ä–æ—Å–∞...")
        import uuid
        session_id = str(uuid.uuid4())
        request = streaming_pb2.StreamRequest(
            prompt="–ü—Ä–∏–≤–µ—Ç, –∫–∞–∫ –¥–µ–ª–∞? –≠—Ç–æ —Ç–µ—Å—Ç–æ–≤—ã–π –∑–∞–ø—Ä–æ—Å –¥–ª—è –ø—Ä–æ–≤–µ—Ä–∫–∏ —Ä–∞–±–æ—Ç—ã gRPC.",
            hardware_id="test-hardware-id-12345",
            session_id=session_id
        )
        print(f"   Session ID: {session_id}")
        print(f"   Prompt: {request.prompt[:50]}...")
        
        try:
            # –ü—Ä–æ–±—É–µ–º –ø–æ–ª—É—á–∏—Ç—å —Ö–æ—Ç—è –±—ã –æ–¥–∏–Ω –æ—Ç–≤–µ—Ç
            response_count = 0
            async for response in stub.StreamAudio(request, timeout=30.0):
                response_count += 1
                if response.HasField('text_chunk'):
                    print(f"‚úÖ –ü–æ–ª—É—á–µ–Ω text_chunk: {response.text_chunk[:50]}...")
                elif response.HasField('audio_chunk'):
                    print(f"‚úÖ –ü–æ–ª—É—á–µ–Ω audio_chunk: {len(response.audio_chunk.audio_data)} –±–∞–π—Ç")
                elif response.HasField('end_message'):
                    print(f"‚úÖ –ü–æ–ª—É—á–µ–Ω end_message: {response.end_message}")
                    break
                elif response.HasField('error_message'):
                    print(f"‚ö†Ô∏è  –ü–æ–ª—É—á–µ–Ω–∞ –æ—à–∏–±–∫–∞ –æ—Ç —Å–µ—Ä–≤–µ—Ä–∞: {response.error_message}")
                    # –≠—Ç–æ –Ω–æ—Ä–º–∞–ª—å–Ω–æ - –∑–∞–ø—Ä–æ—Å –¥–æ—à—ë–ª –¥–æ —Å–µ—Ä–≤–µ—Ä–∞, –∑–Ω–∞—á–∏—Ç gRPC —Ä–∞–±–æ—Ç–∞–µ—Ç!
                    if "session_id" in response.error_message.lower():
                        print("   (–≠—Ç–æ –æ–∂–∏–¥–∞–µ–º–æ - —Å–µ—Ä–≤–µ—Ä —Ç—Ä–µ–±—É–µ—Ç session_id)")
                    break
                
                # –û–≥—Ä–∞–Ω–∏—á–∏–≤–∞–µ–º –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ –æ—Ç–≤–µ—Ç–æ–≤ –¥–ª—è —Ç–µ—Å—Ç–∞
                if response_count >= 5:
                    print("‚ö†Ô∏è –ü–æ–ª—É—á–µ–Ω–æ 5 –æ—Ç–≤–µ—Ç–æ–≤, –ø—Ä–µ—Ä—ã–≤–∞–µ–º —Ç–µ—Å—Ç")
                    break
            
            if response_count == 0:
                print("‚ùå –ù–µ –ø–æ–ª—É—á–µ–Ω–æ –Ω–∏ –æ–¥–Ω–æ–≥–æ –æ—Ç–≤–µ—Ç–∞ –æ—Ç —Å–µ—Ä–≤–µ—Ä–∞")
                print("   –í–æ–∑–º–æ–∂–Ω—ã–µ –ø—Ä–∏—á–∏–Ω—ã:")
                print("   - –°–µ—Ä–≤–µ—Ä –Ω–µ –æ–±—Ä–∞–±–∞—Ç—ã–≤–∞–µ—Ç –∑–∞–ø—Ä–æ—Å—ã")
                print("   - Nginx –Ω–µ –ø—Ä–æ–∫—Å–∏—Ä—É–µ—Ç gRPC –ø—Ä–∞–≤–∏–ª—å–Ω–æ")
                print("   - gRPC —Å–µ—Ä–≤–µ—Ä –Ω–µ –∑–∞–ø—É—â–µ–Ω –Ω–∞ –ø–æ—Ä—Ç—É 50051")
            else:
                print(f"‚úÖ –£—Å–ø–µ—à–Ω–æ –ø–æ–ª—É—á–µ–Ω–æ {response_count} –æ—Ç–≤–µ—Ç–æ–≤")
        
        except grpc.RpcError as e:
            print(f"‚ùå gRPC –æ—à–∏–±–∫–∞: {e.code()} - {e.details()}")
            if e.code() == grpc.StatusCode.UNAVAILABLE:
                print("   –°–µ—Ä–≤–µ—Ä –Ω–µ–¥–æ—Å—Ç—É–ø–µ–Ω. –ü—Ä–æ–≤–µ—Ä—å—Ç–µ:")
                print("   - –ó–∞–ø—É—â–µ–Ω –ª–∏ gRPC —Å–µ—Ä–≤–µ—Ä")
                print("   - –ü—Ä–∞–≤–∏–ª—å–Ω–æ –ª–∏ –Ω–∞—Å—Ç—Ä–æ–µ–Ω Nginx")
                print("   - –î–æ—Å—Ç—É–ø–µ–Ω –ª–∏ –ø–æ—Ä—Ç 443")
            elif e.code() == grpc.StatusCode.DEADLINE_EXCEEDED:
                print("   –¢–∞–π–º–∞—É—Ç. –°–µ—Ä–≤–µ—Ä –Ω–µ –æ—Ç–≤–µ—á–∞–µ—Ç.")
            elif e.code() == grpc.StatusCode.UNIMPLEMENTED:
                print("   –ú–µ—Ç–æ–¥ –Ω–µ —Ä–µ–∞–ª–∏–∑–æ–≤–∞–Ω. –ü—Ä–æ–≤–µ—Ä—å—Ç–µ –≤–µ—Ä—Å–∏—é protobuf.")
        
        finally:
            await channel.close()
            print("‚úÖ –ö–∞–Ω–∞–ª –∑–∞–∫—Ä—ã—Ç")
    
    except Exception as e:
        print(f"‚ùå –ö—Ä–∏—Ç–∏—á–µ—Å–∫–∞—è –æ—à–∏–±–∫–∞: {e}")
        import traceback
        traceback.print_exc()

if __name__ == "__main__":
    host = sys.argv[1] if len(sys.argv) > 1 else "20.63.24.187"
    port = int(sys.argv[2]) if len(sys.argv) > 2 else 443
    
    print("=" * 70)
    print("üîç –¢–ï–°–¢–ò–†–û–í–ê–ù–ò–ï gRPC –ü–û–î–ö–õ–Æ–ß–ï–ù–ò–Ø")
    print("=" * 70)
    print()
    
    asyncio.run(test_grpc_connection(host, port))
